components/Buttons/src/MDCFloatingButton.m

```diff
@@ -72,67 +105,406 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+      return CGSizeMake(MDCFloatingButtonDefaultDimension, MDCFloatingButtonDefaultDimension);
     case MDCFloatingButtonShapeMini:
-      return CGSizeMake([[self class] miniDimension], [[self class] miniDimension]);
-    case MDCFloatingButtonShapeLargeIcon:
-      return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
+      return CGSizeMake(MDCFloatingButtonMiniDimension, MDCFloatingButtonMiniDimension);
   }
 }
 
+- (CGSize)intrinsicContentSizeForModeExpanded {
+  const CGSize intrinsicTitleSize =
+      [self.titleLabel sizeThatFits:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX)];
+  const CGSize intrinsicImageSize =
+      [self.imageView sizeThatFits:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX)];
+  CGFloat intrinsicWidth = intrinsicTitleSize.width + intrinsicImageSize.width +
+      self.imageTitleSpacing + internalLayoutSpacingInsets.left +
+      internalLayoutSpacingInsets.right + self.contentEdgeInsets.left +
+      self.contentEdgeInsets.right;
+  CGFloat intrinsicHeight = MAX(intrinsicTitleSize.height, intrinsicImageSize.height) +
+      self.contentEdgeInsets.top + self.contentEdgeInsets.bottom + internalLayoutSpacingInsets.top +
+      internalLayoutSpacingInsets.bottom;
+  return CGSizeMake(intrinsicWidth, intrinsicHeight);
+}
+
 - (CGSize)intrinsicContentSize {
-  switch (_shape) {
-    case MDCFloatingButtonShapeDefault:
-      return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
-    case MDCFloatingButtonShapeMini:
-      return CGSizeMake([[self class] miniDimension], [[self class] miniDimension]);
-    case MDCFloatingButtonShapeLargeIcon:
-      return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
+  CGSize contentSize = CGSizeZero;
+  if (self.mode == MDCFloatingButtonModeNormal) {
+    contentSize = [self intrinsicContentSizeForModeNormal];
+  } else if (self.mode == MDCFloatingButtonModeExpanded) {
+    contentSize = [self intrinsicContentSizeForModeExpanded];
+  }
+
+  if (self.minimumSize.height > 0) {
+    contentSize.height = MAX(self.minimumSize.height, contentSize.height);
+  }
+  if (self.maximumSize.height > 0) {
+    contentSize.height = MIN(self.maximumSize.height, contentSize.height);
+  }
+  if (self.minimumSize.width > 0) {
+    contentSize.width = MAX(self.minimumSize.width, contentSize.width);
+  }
+  if (self.maximumSize.width > 0) {
+    contentSize.width = MIN(self.maximumSize.width, contentSize.width);
   }
+  return contentSize;
 }
 
 - (void)layoutSubviews {
   // We have to set cornerRadius before laying out subviews so that the boundingPath is correct.
   self.layer.cornerRadius = CGRectGetHeight(self.bounds) / 2;
   [super layoutSubviews];
+
+   if (self.mode == MDCFloatingButtonModeNormal) {
+    return;
+  }
+
+  // Position the imageView and titleView
+  //
+  // +------------------------------------+
+  // |    |  |  |  CEI TOP            |   |
+  // |CEI +--+  |       +-----+       |CEI|
```

Comment:
the CEI is applied after the RGT and LT?  (they applied in series, this image implies they are applied in parallel some how)

Reply:
Sorry, those are "CEI RGT" and "CEI LT", meaning the .left and .right insets of CEI.

Reply:
Created an issue to clean-up the docs: #2566

Reply:
sgtm